File: sdk/typescript/src/state/roomStateMachine.ts
import type { ServerMessage } from "../protocol/messages.ts";
import type { RoomView } from "../protocol/roomViews.ts";
import type { NetplayCloseReason } from "./closeReason.ts";
import type { FrameClockDiagnostics } from "./frameClock.ts";
import { FrameClockTracker } from "./frameClock.ts";
import { HeartbeatTracker, type HeartbeatHealth } from "./heartbeat.ts";
import { PauseCoordinator } from "./pause.ts";
import { ReconnectTokenStore } from "./reconnect.ts";
import { ResyncCoordinator, type NetplayResyncState } from "./resync.ts";

export interface NetplayClientState {
  readonly room: RoomView | null;
  readonly assignedPlayerIndex: number | null;
  readonly latestEventSeq: number;
  readonly roomEpoch: number;
  readonly sessionEpoch: number;
  readonly resync: NetplayResyncState | null;
  readonly runtimeResetRequired: boolean;
  readonly lastError: Extract<NetplayCloseReason, { readonly kind: "relayError" }> | null;
}

export type NetplayEffectivePauseReason =
  | "user"
  | "peer"
  | "connectionRecovery"
  | "stateResync";

export interface NetplayClientDiagnostics {
  readonly assignedPlayerIndex: number | null;
  readonly effectivePauseReason: NetplayEffectivePauseReason | null;
  readonly frameClock: FrameClockDiagnostics;
  readonly heartbeat: HeartbeatHealth;
  readonly heartbeatAck: ReturnType<HeartbeatTracker["lastAck"]>;
  readonly lastError: Extract<NetplayCloseReason, { readonly kind: "relayError" }> | null;
  readonly latestEventSeq: number;
  readonly reconnectTicketAvailable: boolean;
  readonly resync: NetplayResyncState | null;
  readonly roomEpoch: number;
  readonly sessionEpoch: number;
}

export class RoomStateMachine {
  public readonly heartbeat: HeartbeatTracker;
  public readonly frameClock: FrameClockTracker;
  public readonly pause: PauseCoordinator;
  public readonly reconnectTokens: ReconnectTokenStore;
  public readonly resync: ResyncCoordinator;
  public state: NetplayClientState = initialClientState();

  public constructor({
    heartbeat = new HeartbeatTracker(),
    frameClock = new FrameClockTracker(),
    pause = new PauseCoordinator(),
    reconnectTokens = new ReconnectTokenStore(),
    resync = new ResyncCoordinator(),
  }: {
    readonly heartbeat?: HeartbeatTracker;
    readonly frameClock?: FrameClockTracker;
    readonly pause?: PauseCoordinator;
    readonly reconnectTokens?: ReconnectTokenStore;
    readonly resync?: ResyncCoordinator;
  } = {}) {
    this.heartbeat = heartbeat;
    this.frameClock = frameClock;
    this.pause = pause;
    this.reconnectTokens = reconnectTokens;
    this.resync = resync;
  }

  public apply(message: ServerMessage): NetplayClientState {
    if (!this.isMessageCurrent(message)) {
      return this.state;
    }

    switch (message.type) {
      case "roomJoined":
        this.reconnectTokens.applyRoomJoined(message);
        this.updateRoom(message.room, message.yourPlayerIndex);
        break;
      case "roomStateChanged":
      case "compatibilityRequested":
      case "recoveryStarted":
      case "playerReconnected":
      case "playerExited":
      case "inputDelayChanged":
        this.updateRoom(message.room);
        break;
      case "recoveryResyncRequired":
      case "stateHashMismatch":
        this.resync.apply(message, {
          assignedPlayerIndex: this.state.assignedPlayerIndex,
        });
        this.frameClock.reset();
        this.updateRoom(message.room);
        break;
      case "startSession":
        this.resync.markComplete();
        this.resync.clear();
        this.updateRoom(message.room);
        break;
      case "sessionPauseScheduled":
      case "sessionPauseUpdated":
        this.pause.apply(message.pause);
        this.updateRoom(message.room);
        break;
      case "sessionResumeScheduled":
        this.pause.clear(message.sequence);
        this.updateRoom(message.room);
        break;
      case "heartbeatAck":
        this.updateEpochs(message.eventSeq, message.roomEpoch, message.sessionEpoch);
        break;
      case "serverFrame":
        if (this.isRuntimeMessageCurrent(message)) {
          this.frameClock.applyServerFrame(message.frame);
        }
        break;
      case "error":
        this.state = {
          ...this.state,
          lastError: { code: message.code, kind: "relayError", message: message.message },
        };
        break;
      case "inputFrame":
        if (this.isRuntimeMessageCurrent(message)) {
          this.frameClock.markPeerInputFrame(message.input);
        }
        break;
      case "pong":
      case "linkCablePacket":
      case "snapshotChunk":
      case "snapshotComplete":
        break;
    }

    return this.state;
  }

  public reset(): void {
    this.pause.reset();
    this.frameClock.reset();
    this.reconnectTokens.clear();
    this.resync.reset();
    this.state = initialClientState();
  }

  public acknowledgeRuntimeReset(): void {
    this.state = {
      ...this.state,
      runtimeResetRequired: false,
    };
  }

  public isMessageCurrent(message: ServerMessage): boolean {
    if (!("roomEpoch" in message) || !("sessionEpoch" in message)) {
      return true;
    }

    if (this.state.roomEpoch === 0 && this.state.sessionEpoch === 0) {
      return true;
    }

    if (message.roomEpoch < this.state.roomEpoch ||
      message.sessionEpoch < this.state.sessionEpoch) {
      return false;
    }

    return !("eventSeq" in message) ||
      message.roomEpoch !== this.state.roomEpoch ||
      message.sessionEpoch !== this.state.sessionEpoch ||
      message.eventSeq >= this.state.latestEventSeq;
  }

  public isRuntimeMessageCurrent(message: ServerMessage): boolean {
    if (message.type !== "serverFrame") {
      return this.isMessageCurrent(message);
    }

    if (this.state.roomEpoch === 0 && this.state.sessionEpoch === 0) {
      return true;
    }

    return message.frame.roomEpoch === this.state.roomEpoch &&
      message.frame.sessionEpoch === this.state.sessionEpoch;
  }

  public effectivePauseReason(): NetplayEffectivePauseReason | null {
    if (this.resync.currentResync !== null) {
      return this.resync.currentResync.reason === "recovery"
        ? "connectionRecovery"
        : "stateResync";
    }

    const pause = this.pause.currentPause;
    if (pause === null) {
      return null;
    }

    return pause.requestedByPlayerIndex === this.state.assignedPlayerIndex ? "user" : "peer";
  }

  public diagnostics(nowMs: number): NetplayClientDiagnostics {
    return {
      assignedPlayerIndex: this.state.assignedPlayerIndex,
      effectivePauseReason: this.effectivePauseReason(),
      frameClock: this.frameClock.snapshot(),
      heartbeat: this.heartbeat.health(nowMs),
      heartbeatAck: this.heartbeat.lastAck(),
      lastError: this.state.lastError,
      latestEventSeq: this.state.latestEventSeq,
      reconnectTicketAvailable: this.reconnectTokens.current() !== null,
      resync: this.resync.currentResync,
      roomEpoch: this.state.roomEpoch,
      sessionEpoch: this.state.sessionEpoch,
    };
  }

  private updateRoom(room: RoomView, assignedPlayerIndex = this.state.assignedPlayerIndex): void {
    const sessionChanged = this.state.sessionEpoch !== 0 && room.sessionEpoch > this.state.sessionEpoch;
    if (sessionChanged) {
      this.frameClock.reset();
    }

    this.reconnectTokens.updateAcceptedEpoch(room.roomEpoch);
    this.frameClock.applyRoom(room);
    this.state = {
      assignedPlayerIndex,
      lastError: null,
      latestEventSeq: room.eventSeq,
      resync: this.resync.currentResync,
      room,
      roomEpoch: room.roomEpoch,
      runtimeResetRequired: this.state.runtimeResetRequired || sessionChanged,
      sessionEpoch: room.sessionEpoch,
    };
  }

  private updateEpochs(eventSeq: number, roomEpoch: number, sessionEpoch: number): void {
    this.reconnectTokens.updateAcceptedEpoch(roomEpoch);
    this.state = {
      ...this.state,
      latestEventSeq: eventSeq,
      roomEpoch,
      runtimeResetRequired: this.state.runtimeResetRequired ||
        (this.state.sessionEpoch !== 0 && sessionEpoch > this.state.sessionEpoch),
      sessionEpoch,
    };
  }
}

export function initialClientState(): NetplayClientState {
  return {
    assignedPlayerIndex: null,
    lastError: null,
    latestEventSeq: 0,
    resync: null,
    room: null,
    roomEpoch: 0,
    runtimeResetRequired: false,
    sessionEpoch: 0,
  };
}

File: sdk/kotlin/src/main/kotlin/app/shadowboy/netplay/sdk/state/RoomStateMachine.kt
package app.shadowboy.netplay.sdk.state

import app.shadowboy.netplay.sdk.protocol.RoomView
import app.shadowboy.netplay.sdk.protocol.ServerMessage

public data class NetplayClientState(
    public val room: RoomView? = null,
    public val assignedPlayerIndex: Int? = null,
    public val latestEventSeq: Long = 0,
    public val roomEpoch: Long = 0,
    public val sessionEpoch: Long = 0,
    public val resync: NetplayResyncState? = null,
    public val runtimeResetRequired: Boolean = false,
    public val voice: NetplayVoiceGrantState = NetplayVoiceGrantState(),
    public val lastError: NetplayCloseReason.RelayError? = null,
)

public enum class NetplayEffectivePauseReason {
    User,
    Peer,
    ConnectionRecovery,
    StateResync,
}

public data class NetplayClientDiagnostics(
    public val assignedPlayerIndex: Int?,
    public val effectivePauseReason: NetplayEffectivePauseReason?,
    public val frameClock: FrameClockDiagnostics,
    public val heartbeat: HeartbeatHealth,
    public val heartbeatAck: HeartbeatAckState,
    public val lastError: NetplayCloseReason.RelayError?,
    public val latestEventSeq: Long,
    public val reconnectTicketAvailable: Boolean,
    public val resync: NetplayResyncState?,
    public val roomEpoch: Long,
    public val sessionEpoch: Long,
    public val voice: NetplayVoiceDiagnostics,
)

public class RoomStateMachine(
    public val reconnectTokens: ReconnectTokenStore = ReconnectTokenStore(),
    public val pause: PauseCoordinator = PauseCoordinator(),
    public val heartbeat: HeartbeatTracker = HeartbeatTracker(),
    public val frameClock: FrameClockTracker = FrameClockTracker(),
    public val resync: ResyncCoordinator = ResyncCoordinator(),
    public val voice: NetplayVoiceGrantTracker = NetplayVoiceGrantTracker(),
) {
    public var state: NetplayClientState = NetplayClientState()
        private set

    public fun apply(message: ServerMessage): NetplayClientState {
        if (!isMessageCurrent(message)) {
            return state
        }

        when (message) {
            is ServerMessage.RoomJoined -> {
                reconnectTokens.apply(message)
                voice.applyMessage(message)
                updateRoom(message.room, message.yourPlayerIndex)
            }
            is ServerMessage.RoomStateChanged -> updateRoom(message.room)
            is ServerMessage.CompatibilityRequested -> updateRoom(message.room)
            is ServerMessage.RecoveryStarted -> updateRoom(message.room)
            is ServerMessage.PlayerReconnected -> updateRoom(message.room)
            is ServerMessage.PlayerExited -> updateRoom(message.room)
            is ServerMessage.RecoveryResyncRequired -> {
                resync.apply(message, ResyncContext(assignedPlayerIndex = state.assignedPlayerIndex))
                frameClock.reset()
                updateRoom(message.room)
            }
            is ServerMessage.StateHashMismatch -> {
                resync.apply(message, ResyncContext(assignedPlayerIndex = state.assignedPlayerIndex))
                frameClock.reset()
                updateRoom(message.room)
            }
            is ServerMessage.InputDelayChanged -> updateRoom(message.room)
            is ServerMessage.StartSession -> {
                resync.markComplete()
                resync.clear()
                updateRoom(message.room)
            }
            is ServerMessage.SessionPauseScheduled -> {
                pause.apply(message.pause)
                updateRoom(message.room)
            }
            is ServerMessage.SessionPauseUpdated -> {
                pause.apply(message.pause)
                updateRoom(message.room)
            }
            is ServerMessage.SessionResumeScheduled -> {
                pause.clear(message.sequence)
                updateRoom(message.room)
            }
            is ServerMessage.HeartbeatAck -> updateEpochs(
                eventSeq = message.eventSeq,
                roomEpoch = message.roomEpoch,
                sessionEpoch = message.sessionEpoch,
            )
            is ServerMessage.VoiceTokenRefreshed -> {
                voice.applyMessage(message)
                updateEpochs(
                    eventSeq = message.eventSeq,
                    roomEpoch = message.roomEpoch,
                    sessionEpoch = message.sessionEpoch,
                )
            }
            is ServerMessage.Error -> {
                state = state.copy(lastError = NetplayCloseReason.RelayError(message.code, message.message))
            }
            is ServerMessage.InputFrameMessage -> {
                if (isRuntimeMessageCurrent(message)) {
                    frameClock.markPeerInputFrame(message.input)
                }
            }
            is ServerMessage.ServerFrameMessage -> {
                if (isRuntimeMessageCurrent(message)) {
                    frameClock.applyServerFrame(message.frame)
                }
            }
            ServerMessage.Pong,
            is ServerMessage.LinkCablePacketMessage,
            is ServerMessage.SnapshotChunkMessage,
            is ServerMessage.SnapshotComplete -> Unit
        }

        return state
    }

    public fun acknowledgeRuntimeReset() {
        state = state.copy(runtimeResetRequired = false)
    }

    public fun isMessageCurrent(message: ServerMessage): Boolean {
        val roomEpoch = message.roomEpochOrNull ?: return true
        val sessionEpoch = message.sessionEpochOrNull ?: return true

        if (state.roomEpoch == 0L && state.sessionEpoch == 0L) {
            return true
        }

        if (roomEpoch < state.roomEpoch || sessionEpoch < state.sessionEpoch) {
            return false
        }

        val eventSeq = message.eventSeqOrNull
        return eventSeq == null ||
            roomEpoch != state.roomEpoch ||
            sessionEpoch != state.sessionEpoch ||
            eventSeq >= state.latestEventSeq
    }

    public fun isRuntimeMessageCurrent(message: ServerMessage): Boolean =
        when (message) {
            is ServerMessage.ServerFrameMessage -> isExactRuntimeEpochCurrent(
                message.frame.roomEpoch,
                message.frame.sessionEpoch,
            )
            is ServerMessage.InputFrameMessage -> isExactRuntimeEpochCurrent(
                message.roomEpoch,
                message.sessionEpoch,
            )
            is ServerMessage.SnapshotChunkMessage -> isExactRuntimeEpochCurrent(
                message.roomEpoch,
                message.sessionEpoch,
            )
            is ServerMessage.SnapshotComplete -> isExactRuntimeEpochCurrent(
                message.roomEpoch,
                message.sessionEpoch,
            )
            else -> isMessageCurrent(message)
        }

    public fun effectivePauseReason(): NetplayEffectivePauseReason? {
        val currentResync = resync.currentResync
        if (currentResync != null) {
            return if (currentResync.reason == NetplayResyncReason.Recovery) {
                NetplayEffectivePauseReason.ConnectionRecovery
            } else {
                NetplayEffectivePauseReason.StateResync
            }
        }

        val currentPause = pause.currentPause ?: return null
        return if (currentPause.requestedByPlayerIndex == state.assignedPlayerIndex) {
            NetplayEffectivePauseReason.User
        } else {
            NetplayEffectivePauseReason.Peer
        }
    }

    public fun diagnostics(nowMs: Long): NetplayClientDiagnostics =
        NetplayClientDiagnostics(
            assignedPlayerIndex = state.assignedPlayerIndex,
            effectivePauseReason = effectivePauseReason(),
            frameClock = frameClock.snapshot(),
            heartbeat = heartbeat.health(nowMs),
            heartbeatAck = heartbeat.lastAck(),
            lastError = state.lastError,
            latestEventSeq = state.latestEventSeq,
            reconnectTicketAvailable = reconnectTokens.current() != null,
            resync = resync.currentResync,
            roomEpoch = state.roomEpoch,
            sessionEpoch = state.sessionEpoch,
            voice = voice.diagnostics(),
        )

    public fun reset() {
        reconnectTokens.clear()
        pause.reset()
        frameClock.reset()
        resync.reset()
        voice.reset()
        state = NetplayClientState()
    }

    private fun updateRoom(room: RoomView, assignedPlayerIndex: Int? = state.assignedPlayerIndex) {
        val sessionChanged = state.sessionEpoch != 0L && room.sessionEpoch > state.sessionEpoch
        if (sessionChanged) {
            frameClock.reset()
        }

        reconnectTokens.updateAcceptedEpoch(room.roomEpoch)
        voice.applyRoom(room)
        frameClock.applyRoom(room)
        state = state.copy(
            room = room,
            assignedPlayerIndex = assignedPlayerIndex,
            latestEventSeq = room.eventSeq,
            roomEpoch = room.roomEpoch,
            sessionEpoch = room.sessionEpoch,
            resync = resync.currentResync,
            runtimeResetRequired = state.runtimeResetRequired || sessionChanged,
            voice = voice.state,
            lastError = null,
        )
    }

    private fun updateEpochs(eventSeq: Long, roomEpoch: Long, sessionEpoch: Long) {
        reconnectTokens.updateAcceptedEpoch(roomEpoch)
        state = state.copy(
            latestEventSeq = eventSeq,
            roomEpoch = roomEpoch,
            sessionEpoch = sessionEpoch,
            voice = voice.state,
            runtimeResetRequired = state.runtimeResetRequired ||
                (state.sessionEpoch != 0L && sessionEpoch > state.sessionEpoch),
        )
    }

    private fun isExactRuntimeEpochCurrent(roomEpoch: Long, sessionEpoch: Long): Boolean {
        if (state.roomEpoch == 0L && state.sessionEpoch == 0L) {
            return true
        }

        return roomEpoch == state.roomEpoch && sessionEpoch == state.sessionEpoch
    }
}
